tower/src/retry/backoff.rs

```diff
@@ -0,0 +1,228 @@
+//! This module contains generic backoff utlities to be used with the retry
+//! layer.
+//!
+//! The [`Backoff`] trait is a generic way to represent backoffs that can use
+//! any timer type.
+//!
+//! [`ExponentialBackoff`] implements [`Backoff`] and provides a batteries
+//! included exponential backoff and jitter strategy.
+
+use rand::Rng;
+use rand::{rngs::SmallRng, thread_rng, SeedableRng};
+use std::fmt::Display;
+use std::future::Future;
+use std::time::Duration;
+use tokio::time;
+
+/// A backoff trait where a single mutable refernces represents a single
+/// backoff session. Implementors must also implement [`Clone`] which will
+/// reset the backoff back to the default state for the next session.
```

Comment:
`Clone` resets it to the default? That seems different than normal `Clone` behavior, and easy to get wrong. If I use a backoff a few times, and it's up to 2150ms, a normal clone with just give me the same thing at the same time.

Reply:
Yeah, I originally had a `reset(&self) -> Self` which I could add back instead of  using clone. What do you think? or we could have a separate kinda factory trait. What do you think?

Reply:
To stay object-safe, a reset function could just mutate self to 0.

Reply:
The issue is that we need to basically have a factory to get owned copies of a backoff type that we can pass into the future. So we need some sort of way to share that. We could go with some sort of factory type but that is two traits which is more a pita. We could try `Stream` for this though.

Reply:
Could still ask that backoffs be clone...

Another option is to make the `next_backoff` return `Self`. It well encodes the meaning, but makes the trait object safe.

Reply:
IMO, a `Backoff` trait which returns a `Stream` for each retry session is a cleaner design than one where cloning the `Backoff` type resets the backoff.

This would also allow more flexible user implementations; for example, a `tokio::time::Interval` could be returned as a backoff using the `tokio-stream` wrapper. Also, stream combinators could be used to modify an existing backoff: for example, if I wanted to do something every time a backoff completes (such as incrementing metrics), I could wrap an existing backoff implementation's return value in `Stream::inspect` or similar.

Reply:
I agree that's cleaner. But then we have to depend on `Stream`, which... I don't really want...

Reply:
My issue with stream is the requirement to implement stuff with `poll` which I want to avoid when you just need to work with Delay + Duration types which are much nicer to use than `poll` etc.

Reply:
Here is a proposal I have:

```rust

/// A backoff maker that can produce backoff types for retries.
pub trait MakeBackoff {
    /// Backoff returned from this maker.
    type Backoff: Backoff;

    /// Make a new backoff type that can be used for a single request
    /// retry strategy.
    fn make_backoff(&mut self) -> Self::Backoff;
}

/// A backoff trait where a single mutable reference represents a single
/// backoff session. Implementors must also implement [`Clone`] which will
/// reset the backoff back to the default state for the next session.
pub trait Backoff {
    /// The future associated with each backoff. This usually will be some sort
    /// of timer.
    type Future: Future<Output = ()>;

    /// Initiate the next backoff in the sequence.
    fn next_backoff(&mut self) -> Self::Future;
}
```

Reply:
https://github.com/tower-rs/tower/pull/685/commits/26fd2f8b7554aa9993216c874cd34c663af02eb2 adds this new trait